### PR TITLE
Fix reinterpretAs*() on big endian machines

### PR DESCRIPTION
**Pull request by suzyw-w**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix reinterpretAs*() on big endian machines (e.g. s390x).

### CODE REVIEW

File: src/Functions/reinterpretAs.cpp [hunk 1]
```diff
@@ -176,18 +176,23 @@ class FunctionReinterpret : public IFunction
                     size_t size = offsets_from.size();
                     auto & vec_res = col_res->getData();
                     vec_res.resize_fill(size);
+                    size_t copy_size = 0;
```
Comment (rschu1ze) [inline src/Functions/reinterpretAs.cpp:179]: No need to declare `copy_size` and `offset_to` outside the loop, they are both stack variables and assigning something to them isn't performance critical. Feel free to do their definition inside the loop.

File: src/Functions/reinterpretAs.cpp [hunk 2]
```diff
@@ -176,18 +176,23 @@ class FunctionReinterpret : public IFunction
...
+                    size_t offset_to = 0;
 
                     size_t offset = 0;
                     for (size_t i = 0; i < size; ++i)
                     {
+                        copy_size = std::min(static_cast<UInt64>(sizeof(ToFieldType)), offsets_from[i] - offset - 1);
+                        offset_to = sizeof(ToFieldType) > copy_size ? sizeof(ToFieldType) - copy_size : 0;
```
Comment (rschu1ze) [inline src/Functions/reinterpretAs.cpp:186]: Let's move this calculation into the `else` case (l. 191.)

File: src/Functions/reinterpretAs.cpp [hunk 3]
```diff
@@ -176,18 +176,23 @@ class FunctionReinterpret : public IFunction
...
                                 &data_from[offset],
-                                std::min(static_cast<UInt64>(sizeof(ToFieldType)), offsets_from[i] - offset - 1));
+                                copy_size);
                         else
-                            reverseMemcpy(&vec_res[i],
+                            reverseMemcpy(
+                                reinterpret_cast<char*>(&vec_res[i]) + offset_to,
                                 &data_from[offset],
-                                std::min(static_cast<UInt64>(sizeof(ToFieldType)), offsets_from[i] - offset - 1));
+                                copy_size);
```
Comment (rschu1ze) [inline src/Functions/reinterpretAs.cpp:195]: About l. 203: Below, I left a comment about similar changes potentially necessary for FixedString. The same would apply here - maybe you would like to check, thanks.

File: src/Functions/reinterpretAs.cpp [hunk 4]
```diff
@@ -242,9 +247,15 @@ class FunctionReinterpret : public IFunction
                     to.resize_fill(size);
 
                     static constexpr size_t copy_size = std::min(sizeof(From), sizeof(To));
+                    size_t offset_to = sizeof(To) > sizeof(From) ? sizeof(To) - sizeof(From):0;
```
Comment (rschu1ze) [inline src/Functions/reinterpretAs.cpp:250]: Let's move that into the `else` case, l. 257.

File: src/Functions/reinterpretAs.cpp [hunk 5]
```diff
@@ -322,11 +333,21 @@ class FunctionReinterpret : public IFunction
             StringRef data = src.getDataAt(i);
 
             /// Cut trailing zero bytes.
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+            size_t index = 0;
+            while (index < data.size && data.data[index] == 0)
+                    index++;
```
[the remark this answers is not in the document]
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp]: Updated

File: src/Functions/reinterpretAs.cpp [hunk 6]
```diff
@@ -322,11 +333,21 @@ class FunctionReinterpret : public IFunction
             StringRef data = src.getDataAt(i);
 
             /// Cut trailing zero bytes.
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
```
[the remark this answers is not in the document]
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp]: Updated

File: src/Functions/reinterpretAs.cpp [hunk 7]
```diff
@@ -322,11 +333,21 @@ class FunctionReinterpret : public IFunction
...
             while (data.size && data.data[data.size - 1] == 0)
                 --data.size;
-
+#endif
             data_to.resize(offset + data.size + 1);
-            memcpy(&data_to[offset], data.data, data.size);
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+            reverseMemcpy(&data_to[offset], data.data+index, data.size);
+#else
+            memcpy(&data_to[offset],data.data, data.size);
```
[the remark this answers is not in the document]
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp]: Updated

File: src/Functions/reinterpretAs.cpp [hunk 8]
```diff
@@ -322,11 +333,21 @@ class FunctionReinterpret : public IFunction
...
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+            size_t index = 0;
+            while (index < data.size && data.data[index] == 0)
+                    index++;
+            data.size -= index;
+#else
             while (data.size && data.data[data.size - 1] == 0)
                 --data.size;
-
+#endif
```
[the remark this answers is not in the document]
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp:351]: Include fix for reinterpretAsfixedString

File: src/Functions/reinterpretAs.cpp [hunk 9]
```diff
@@ -322,11 +333,21 @@ class FunctionReinterpret : public IFunction
...
             data_to.resize(offset + data.size + 1);
-            memcpy(&data_to[offset], data.data, data.size);
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+            reverseMemcpy(&data_to[offset], data.data+index, data.size);
+#else
+            memcpy(&data_to[offset],data.data, data.size);
+#endif
```
[the remark this answers is not in the document]
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp:357]: Removed

File: src/Functions/reinterpretAs.cpp (hunk 1 above)
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp]: Updated

File: src/Functions/reinterpretAs.cpp (hunk 2 above)
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp]: Updated

File: src/Functions/reinterpretAs.cpp (hunk 4 above)
Comment (SuzyWangIBMer) [reply src/Functions/reinterpretAs.cpp]: Updated

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit ce471a2e8bc9572c19b5fc2ab1bf92867c4ffe82 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49198/ce471a2e8bc9572c19b5fc2ab1bf92867c4ffe82/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by rschu1ze — COMMENTED** [review]
Lgtm overall, only the LE build still fails.

**Comment by rschu1ze** [comment]
Could you merge from master once more, thanks.

File: src/Functions/reinterpretAs.cpp [hunk 10]
```diff
@@ -216,7 +222,13 @@ class FunctionReinterpret : public IFunction
 
                     for (size_t i = 0; i < size; ++i)
                     {
-                        memcpy(&vec_res[i], &data_from[offset], copy_size);
+                        if constexpr (std::endian::native == std::endian::little)
+                            memcpy(&vec_res[i], &data_from[offset], copy_size);
+                        else
+                        {
+                            size_t offset_to = sizeof(ToFieldType) > copy_size ? sizeof(ToFieldType) - copy_size : 0;
+                            memcpy(reinterpret_cast<char*>(&vec_res[i]) + offset_to, &data_from[index - offset], copy_size);
```
Comment (azat) [inline src/Functions/reinterpretAs.cpp:230]: Apparently this should also use `reverseMemcpy` as the case for `String` - https://github.com/ClickHouse/ClickHouse/pull/76253

**Pull request merged**